bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/finders/MDNSAddonSuggestionFinder.java

```diff
@@ -0,0 +1,125 @@
+
+    private final Logger logger = LoggerFactory.getLogger(MDNSAddonSuggestionFinder.class);
+    private final Map<String, ServiceInfo> services = new ConcurrentHashMap<>();
+    private final MDNSClient mdnsClient;
+
+    @Activate
+    public MDNSAddonSuggestionFinder(@Reference MDNSClient mdnsClient) {
+        this.mdnsClient = mdnsClient;
+    }
+
+    public void addService(@Nullable ServiceInfo service) {
+        if (service != null) {
+            String qualifiedName = service.getQualifiedName();
+            if (services.put(qualifiedName, service) == null && logger.isTraceEnabled()) {
+                logger.trace("mDNS service name={}, properties={}", qualifiedName,
+                        Collections.list(service.getPropertyNames()).stream()
+                                .map(n -> n + "=" + service.getPropertyString(n)).toList());
+            }
+        }
+    }
+
+    @Deactivate
+    public void close() {
+        services.clear();
+        super.close();
+    }
+
+    @Override
+    public Set<AddonInfo> getSuggestedAddons() {
+        Set<AddonInfo> result = new HashSet<>();
+        addonCandidates.forEach(candidate -> {
+            candidate.getDiscoveryMethods().stream().filter(method -> SERVICE_TYPE.equals(method.getServiceType()))
+                    .forEach(method -> {
+                        Map<String, Pattern> map = method.getMatchProperties().stream().collect(
+                                Collectors.toMap(property -> property.getName(), property -> property.getPattern()));
+                        services.values().stream().forEach(service -> {
+                            if (method.getMdnsServiceType().equals(service.getType())
+                                    && propertyMatches(map, NAME, service.getName())
+                                    && Collections.list(service.getPropertyNames()).stream().allMatch(
+                                            name -> propertyMatches(map, name, service.getPropertyString(name)))) {
+                                result.add(candidate);
+                                logger.debug("Addon '{}' will be suggested (via mDNS)", candidate.getUID());
+                            }
+                        });
+                    });
+        });
+        return result;
+    }
+
+    @Override
+    public void serviceAdded(@Nullable ServiceEvent event) {
+        if (event != null) {
+            addService(event.getInfo());
+        }
+    }
+
+    @Override
+    public void serviceRemoved(@Nullable ServiceEvent event) {
+    }
+
+    @Override
+    public void serviceResolved(@Nullable ServiceEvent event) {
+        if (event != null) {
+            addService(event.getInfo());
+        }
+    }
+
+    @Override
+    public void setAddonCandidates(List<AddonInfo> candidates) {
+        super.setAddonCandidates(candidates);
+        candidates.forEach(c -> c.getDiscoveryMethods().stream().filter(m -> SERVICE_TYPE.equals(m.getServiceType()))
```

Comment:
Is there no other way than starting to listen when we first call the suggestions REST endpoint? It looks like that may force us to call the endpoint twice to get all.

Reply:
My idea was to call setAddonCandidates whenever the framework loads an AddonInfoProvider component so the search would start prior to the rest api call. But see my response above.

Reply:
> My idea was to call setAddonCandidates whenever the framework loads an AddonInfoProvider component so the search would start prior to the rest api call. But see my response above.

That's also what I would prefer. Unfortunately, the AddonInfoRegistry does not have a list of listeners to be updated when an AddonInfoProvider is added/removed. You keep your own reference to AddonInfoProviders, but that somewhat defeats the purpose of the registry I would think.

Reply:
> would be an easy way to know if an AddonInfoProvider was added.
> You could add .. to the BaseAddonSuggestionFinder so it gets updated with new AddonInfoProviders.

Umm. IMHO that is exactly what I am doing in the AddonSuggestionFinderService (which you complained about above). And the AddonSuggestionFinderService passes that candidate info to each of its AddonSuggestionFinder class instances.

Reply:
> Is there no other way than starting to listen when we first call the suggestions REST endpoint?

PS IMHO that method is NOT called when the endpoint is called. It is called when a new AddonInfoProvider is loaded.

Reply:
> Umm. IMHO that is exactly what I am doing in the AddonSuggestionFinderService (which you complained about above). And the AddonSuggestionFinderService passes that candidate info to each of its AddonSuggestionFinder class instances.

Yes, sorry, you are right. It should get updated. I need to look deeper. It doesn't explain why I don't get all results at the first call of the REST API, but only on the second call.

Reply:
I see the problem now. The service caches the AddonInfoProviders immediately they are instantiated, but it only queries them when the rest call is made. I will make a change so that the query is started immediately.